library/kani/src/slice.rs

```diff
@@ -43,45 +44,57 @@ fn any_range<const LENGTH: usize>() -> (usize, usize) {
 /// # Example:
 ///
 /// ```rust
-/// let slice: kani::slice::NonDetSlice<u8, 5> = kani::slice::any_slice();
+/// let slice: kani::slice::AnySlice<u8, 5> = kani::slice::any_slice();
 /// foo(&slice); // where foo is a function that takes a slice and verifies a property about it
 /// ```
-pub struct NonDetSlice<T, const MAX_SLICE_LENGTH: usize> {
-    arr: [T; MAX_SLICE_LENGTH],
+pub struct AnySlice<T, const MAX_SLICE_LENGTH: usize> {
+    layout: Layout,
+    ptr: *mut T,
     slice_len: usize,
 }
 
-impl<T, const MAX_SLICE_LENGTH: usize> NonDetSlice<T, MAX_SLICE_LENGTH> {
+impl<T, const MAX_SLICE_LENGTH: usize> AnySlice<T, MAX_SLICE_LENGTH> {
     fn new() -> Self {
-        let arr: [T; MAX_SLICE_LENGTH] = unsafe { any_raw() };
         let slice_len: usize = any();
         assume(slice_len <= MAX_SLICE_LENGTH);
-        Self { arr, slice_len }
+        let layout = Layout::array::<T>(slice_len).unwrap();
+        let ptr = unsafe { alloc(layout) };
+        Self { layout, ptr: ptr as *mut T, slice_len }
     }
 
     pub fn get_slice(&self) -> &[T] {
-        &self.arr[..self.slice_len]
+        unsafe { std::slice::from_raw_parts(self.ptr, self.slice_len) }
     }
 
     pub fn get_slice_mut(&mut self) -> &mut [T] {
-        &mut self.arr[..self.slice_len]
+        unsafe { std::slice::from_raw_parts_mut(self.ptr, self.slice_len) }
     }
 }
 
-impl<T, const MAX_SLICE_LENGTH: usize> Deref for NonDetSlice<T, MAX_SLICE_LENGTH> {
+impl<T, const MAX_SLICE_LENGTH: usize> Drop for AnySlice<T, MAX_SLICE_LENGTH> {
+    fn drop(&mut self) {
+        if self.slice_len > 0 {
+            unsafe {
+                dealloc(self.ptr as *mut u8, self.layout);
+            }
+        }
+    }
+}
+
+impl<T, const MAX_SLICE_LENGTH: usize> Deref for AnySlice<T, MAX_SLICE_LENGTH> {
     type Target = [T];
 
     fn deref(&self) -> &Self::Target {
         self.get_slice()
     }
 }
 
-impl<T, const MAX_SLICE_LENGTH: usize> DerefMut for NonDetSlice<T, MAX_SLICE_LENGTH> {
+impl<T, const MAX_SLICE_LENGTH: usize> DerefMut for AnySlice<T, MAX_SLICE_LENGTH> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         self.get_slice_mut()
     }
 }
 
-pub fn any_slice<T, const MAX_SLICE_LENGTH: usize>() -> NonDetSlice<T, MAX_SLICE_LENGTH> {
-    NonDetSlice::<T, MAX_SLICE_LENGTH>::new()
+pub fn any_slice<T, const MAX_SLICE_LENGTH: usize>() -> AnySlice<T, MAX_SLICE_LENGTH> {
```

Comment:
Just wanna make sure you've seen my comments about making this method unsafe or using the Arbitrary trait.

Reply:
Yes, I have seen it, but I haven't addressed it yet. I think we have discussed this before, but I don't recall what we concluded. What I'm thinking right now is to perhaps add `Invariant` as a trait bound for `T` in `AnySlice` and set every element in the slice to `kani::any()`. We can also provide an unsafe version, e.g. `AnyRawSlice` that does not require `Invariant` for `T`. What do you think?

Reply:
As discussed offline, changed `T` to require `Arbitrary`, and added an unsafe `any_raw_slice` version that doesn't require `Arbitrary`.